ch06ex02/test/LoginTest.js

```diff
@@ -38,40 +43,47 @@ describe("Login", function () {
     // NOTE: We're appending some characters to the password here to give us an invalid password to test the messaging.
-    await driver.findElement({ id: "password" }).sendKeys(PASSWORD + "pants");
+    await driver.findElement({ id: "password" }).sendKeys(PASSWORD + Math.random());
     await driver.findElement({ css: "button" }).click();
 
-    assert(
-      await driver.findElement({ css: ".flash.error" }).isDisplayed(),
-      "Success message not displayed"
-    );
+    // NOTE: Another suboptimal issue might be the fact we're doing the DOM
+    // lookup of `driver.findElement({css:".flash.error"})` at least 3x in
+    // this test case. Might be worth seeing if we can do it once and cache
+    // the potentially expensive lookup or if that has downstream issues.
+    const el = await driver.findElement({ css: ".flash.error" });
+
+    assert(await el.isDisplayed(), "Success message not displayed");
+
+    // OPTION 1: Use `assert.equal()` to compare the full error text.
     assert.equal(
       // NOTE: Need to wrap the `.getText()` bit in parens since it returns a
       // promise and we want to resolve said promise before calling the
       // string's `.trim()` method. Perhaps a better way, but I wanted to use
       // `assert.equal()` here for the superior error diffing.
-      (await driver.findElement({ css: ".flash.error" }).getText()).trim(),
+      (await el.getText()).trim(),
       // This is goofy because the `getText()` includes a newline and the `×`
       // icon for dismissing the flash message.
       "Your password is invalid!\n×",
       "Invalid password"
     );
 
+    // OPTION 2: Use vanilla `assert()` and `<string>.includes()` to check partial text.
     assert(
       // Perhaps slightly better string checking method using `.includes()`
       // to only check the substring we care about. We still need to wrap
       // the `await ....getText()` call in parens to resolve the promise
       // before we can call `.includes()`. Error messaging won't be as nice
       // if the test fails but probably a good enough compromise.
-      (await driver.findElement({ css: ".flash.error" }).getText()).includes(
-        "Your password is invalid!"
-      ),
+      (await el.getText()).includes("Your password is invalid!"),
       "Invalid password"
     );
 
-    // NOTE: Another suboptimal issue might be the fact we're doing the DOM
-    // lookup of `driver.findElement({css:".flash.error"})` at least 3x in
-    // this test case. Might be worth seeing if we can do it once and cache
-    // the potentially expensive lookup or if that has downstream issues.
+    // OPTION 3: Use `assert.match()` to use regexp string matching to check the invalid password text.
+    // https://nodejs.org/api/assert.html#assert_assert_match_string_regexp_message
+    assert.match(
+      await el.getText(),
+      /^Your password is invalid!/,
+      "Invalid password"
+    );
```

Comment:
Add RegExp string checking as OPTION 3.
This feels like the best option so far, but probably moot since we'll be swapping out assert engines in the future.